### PR TITLE
Add metrics for SpamAssassin and Tika

### PR DESCRIPTION
**Pull request by chibenwa**
We suspect high delays upon message reception and message moving to be due to Tika and SpamAssassin that we recently added to our set up.

I order to confirm such a hypothesis, we need additional metrics on these two components. I added this as part of this pull request.

### CODE REVIEW

File: mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java [hunk 1]
```diff
@@ -71,8 +75,12 @@ private ObjectMapper initializeObjectMapper() {
 
     @Override
     public ParsedContent extractContent(InputStream inputStream, String contentType) throws Exception {
-        ContentAndMetadata contentAndMetadata = convert(tikaHttpClient.recursiveMetaDataAsJson(inputStream, contentType));
-        return new ParsedContent(contentAndMetadata.getContent(), contentAndMetadata.getMetadata());
+        return metricFactory.withMetric("tikaTextExtraction", Throwing.supplier(
+            () -> {
```
Comment (mbaechler) [inline mailbox/tika/src/main/java/org/apache/james/mailbox/tika/TikaTextExtractor.java]: please extract to a new method when your lambda is that long

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java [hunk 2]
```diff
@@ -52,19 +54,20 @@
...
+    private final MetricFactory metricFactory;
     private final String spamdHost;
-
     private final int spamdPort;
 
     /**
      * Init the spamassassin invoker
-     * 
+     *
+     * @param metricFactory
```
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: please don't add useless doc

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java [hunk 3]
```diff
@@ -202,17 +184,25 @@ public boolean learnAsSpam(InputStream message, String user) throws MessagingExc
...
      */
     public boolean learnAsHam(InputStream message, String user) throws MessagingException {
+        return metricFactory.withMetric(
+            "spamAssassin-ham-report",
+            Throwing.supplier(
+                () -> performReport(message, user, "ham"))
+                .sneakyThrow());
+    }
+
+    public boolean performReport(InputStream message, String user, String messageClass) throws MessagingException {
```
Comment (aduprat) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: This method name is not really clear, IMO. Maybe `learn`

and also it should be private.

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java [hunk 4]
```diff
@@ -84,7 +87,11 @@ public SpamAssassinResult scanMail(MimeMessage message, String user) throws Mess
     }
 
     public SpamAssassinResult scanMail(MimeMessage message) throws MessagingException {
-        return scanMailWithAdditionalHeaders(message);
+        return metricFactory.withMetric(
+            "spamAssassin-check",
+            Throwing.supplier(
+                () -> scanMailWithAdditionalHeaders(message))
```
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: you don't use additional headers but using this vararg method prevent you from using method reference.
could you create a method that doesn't take additional headers ?

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java (hunk 3 above)
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: do you need it to be public ?

File: server/container/guice/mailbox-plugin-spamassassin/src/main/java/org/apache/james/modules/spamassassin/SpamAssassinListenerModule.java [hunk 5]
```diff
@@ -58,24 +56,19 @@ protected void configure() {
...
+                                                          StoreMailboxManager storeMailboxManager) {
+            this.spamAssassinListener = spamAssassinListener;
             this.storeMailboxManager = storeMailboxManager;
-            this.mapperFactory = mapperFactory;
         }
 
         @Override
         public void initModule() {
             try {
-                SpamAssassinListener spamAssassinListener = new SpamAssassinListener(new SpamAssassin(spamAssassinConfiguration), mapperFactory);
```
Comment (aduprat) [inline server/container/guice/mailbox-plugin-spamassassin/src/main/java/org/apache/james/modules/spamassassin/SpamAssassinListenerModule.java]: Then we should declare it with a `Scopes.SINGLETON` scope, no?

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java [hunk 6]
```diff
@@ -202,17 +184,25 @@ public boolean learnAsSpam(InputStream message, String user) throws MessagingExc
      *             if an error occured during learning.
      */
     public boolean learnAsHam(InputStream message, String user) throws MessagingException {
+        return metricFactory.withMetric(
+            "spamAssassin-ham-report",
+            Throwing.supplier(
+                () -> performReport(message, user, "ham"))
```
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: please don't use a string for something as important as SPAM and HAM values

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java (hunk 3 above)
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: when reading this method name, I don't really understand what it does. Would you want to rename it to "reportMessageAs" or something similar ?

Comment (chibenwa) [reply server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: I prefer matthieu **reportMessageAs** proposal.

File: server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java (hunk 4 above)
Comment (chibenwa) [reply server/container/util/src/main/java/org/apache/james/util/scanner/SpamAssassinInvoker.java]: > could you create a method that doesn't take additional headers ?

Yes sure. It makes things more obvious.

>  using this vararg method prevent you from using method reference.

Not only. It's a Supplier, so even with that change you can not use method reference.

**Comment by chibenwa** [comment]
```
[de83046826e24cc6022717908dd361e7a84286b5] Failed tests: 
[de83046826e24cc6022717908dd361e7a84286b5]   JPAJamesServerTest.jpaGuiceServerShouldUpdateQuota:88 expected:<...ames.local (STORAGE [12] 50)
[de83046826e24cc6022717908dd361e7a84286b5] AAAB OK GETQUO...> but was:<...ames.local (STORAGE [0] 50)
[de83046826e24cc6022717908dd361e7a84286b5] AAAB OK GETQUO...>
```

test this please

**Comment by mbaechler** [comment]
merged

**Pull request closed without merging**